src/dtree.ml

```diff
@@ -0,0 +1,200 @@
+open Terms
+open Extra
+
+(** Type of the leaves of the tree.  See {!file:terms.ml}, {!recfield:rhs}. *)
+type action = (term_env, term) Bindlib.mbinder
+
+(** Tree type.  {!const:Leaf}[(t, a)] are the leaves of the tree holding the
+    targets of the rewriting as an {!type:action}; for [t], see
+    {!recfield:switch} of {!type:node_data}.  The term option evinces--if the
+    leaf has been obtained by a switch--the term that has been matched before
+    reaching the leaf.  The {!const:Node}[n] constructor allows primarily to
+    perform a switch among subtrees recorded in {!recfield:children} of [n].
+    {!const:Fail} is a matching failure. *)
+type t = Leaf of term option * action
+       | Node of node_data
+       | Fail
+
+(** Data contained in a node of the tree.  {!recfield:switch} contains the
+    term on which the switch that gave birth to this node has been performed
+    (or none if the node has been obtained by a default matrix);
+    {!recfield:swap} indicates whether the columns of the matrix have been
+    swapped before the switch, with an int indicating the column which has
+    been swapped with the first one (or None if no swap is performed) and
+    {!recfield:children} contains the subtrees. *)
+and node_data = { switch : term option
+                ; swap : int option
+                ; children : t list}
+
+(** [iter l n f t] is a generic iterator on trees; with function [l] performed
+    on leaves, function [n] performed on nodes, [f] returned in case of
+    {!const:Fail} on tree [t]. *)
+let iter : (term option -> action -> 'a) ->
+  (term option -> int option -> t list -> 'a) ->
+  'a -> t -> 'a = fun do_leaf do_node fail t ->
+  let rec loop = function
+    | Leaf(teo, a)                                   -> do_leaf teo a
+    | Fail                                           -> fail
+    | Node({ switch = s ; swap = p ; children = c }) ->
+      do_node s p (List.map loop c) in
+  loop t
+
+(** [to_dot f t] creates a dot graphviz file [f].gv for tree [t]. *)
```

Comment:
Please, explain the semantics of your output tree. It is not clear to me. And add an example. Thank you.

Reply:
more in 02ec495